codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+                    rustTemplate(
+                        """
+                        let mut tokens_owned = #{json_token_iter}(input).peekable();
+                        let tokens = &mut tokens_owned;
+                        #{expect_start_object}(tokens.next())?;
+                        """,
+                        *codegenScope
+                    )
+                    deserializeStructInner(outputShape, httpDocumentMembers)
+                    expectEndOfTokenStream()
+                }
+                rust("Ok(builder)")
+            }
+        }
+    }
+
+    override fun errorParser(errorShape: StructureShape): RuntimeType? {
+        if (errorShape.members().isEmpty()) {
+            return null
+        }
+        val fnName = symbolProvider.deserializeFunctionName(errorShape) + "json_err"
+        return RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                "pub fn $fnName(input: &[u8], mut builder: #{Builder}) -> Result<#{Builder}, #{Error}>",
+                "Builder" to errorShape.builderSymbol(symbolProvider),
+                *codegenScope
+            ) {
+                rustTemplate(
+                    """
+                    let mut tokens_owned = #{json_token_iter}(input).peekable();
+                    let tokens = &mut tokens_owned;
+                    #{expect_start_object}(tokens.next())?;
+                    """,
+                    *codegenScope
+                )
+                deserializeStructInner(errorShape, errorShape.members())
+                expectEndOfTokenStream()
+                rust("Ok(builder)")
+            }
+        }
+    }
+
+    override fun documentParser(operationShape: OperationShape): RuntimeType {
+        val fnName = "parse_document"
+        return RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                "pub fn $fnName(input: &[u8]) -> Result<#{Document}, #{Error}>",
+                "Document" to RuntimeType.Document(runtimeConfig),
+                *codegenScope,
+            ) {
+                rustTemplate(
+                    """
+                    let mut tokens_owned = #{json_token_iter}(input).peekable();
+                    let tokens = &mut tokens_owned;
+                    """,
+                    *codegenScope
+                )
+                rustTemplate("let result = #{expect_document}(tokens);", *codegenScope)
+                expectEndOfTokenStream()
+                rust("result")
+            }
+        }
+    }
+
+    private fun RustWriter.expectEndOfTokenStream() {
+        rustBlock("if let Some(_) = tokens.next()") {
+            rustTemplate(
+                "return Err(#{Error}::custom(\"found more JSON tokens after completing parsing\"));",
+                *codegenScope
+            )
+        }
+    }
+
+    private fun RustWriter.deserializeStructInner(shape: StructureShape, includedMembers: Collection<MemberShape>? = null) {
```

Comment:
you don't actually need `shape` here—just pass in the list of members. Callers can specify `shape.members()`